test/clustertest/tests/ClusterUpgradeTest.cpp

```diff
@@ -0,0 +1,172 @@
+
+        // We need the tag names from each release.
+        array<string, RECENT_RELEASES_TO_CHECK> tagNames;
+        for (size_t i = 0; i < RECENT_RELEASES_TO_CHECK; i++) {
+            STable j2 = SParseJSONObject(j1.front());
+            auto tag = j2.find("tag_name");
+            if (tag != j2.end()) {
+                tagNames[i] = tag->second;
+            }
+            if (j1.size()) {
+                j1.pop_front();
+            }
+        }
+
+        // Now choose the one to use. We want to test against the msot recent release that isn't the commit we're currently on.
+        // the commit number of the tag: git rev-list -n 1 $TAG
+        // The commit number we're currently on: git rev-parse HEAD
+        // If the current commit matches the tested tag, the script returns 1 and we check the next one. When the script returns 0, that's the release we'll use.
+        for (const auto& tagName : tagNames) {
+            string checkIfOnLatestTag = "/bin/bash -c 'if [ \"`git rev-list -n 1 " + tagName + "`\" = \"`git rev-parse HEAD`\" ]; then exit 1; else exit 0; fi'";
+            int result = system(checkIfOnLatestTag.c_str());
+            if (!result) {
+                bedrockTagName = tagName;
+                break;
+            }
+        }
+
+        // Make sure we got something to test.
+        ASSERT_NOT_EQUAL(bedrockTagName, "");
+
+        // If you'd like to test against a particular tag, uncomment the following line. The value chosen here was a
+        // known bad version that failed to escalate commands at upgrade when first deployed.
+        // bedrockTagName = "2022-05-06";
+
+        // If we've already built this, don't bother doing it again. This makes running this test multiple times in a
+        // row much faster.
+        prodBedrockName = "/tmp/bedrock-" + bedrockTagName; 
+        if (!SFileExists(prodBedrockName)) {
+            // Get a directory we can work in.
+            char brReleaseDirArr[] = "/tmp/br-prod-test-XXXXXX";
+            ASSERT_EQUAL(mkdtemp(brReleaseDirArr), brReleaseDirArr);
+            string brReleaseDir(brReleaseDirArr, sizeof(brReleaseDirArr) - 1);
+
+            // Clone bedrock.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && git clone https://github.com/Expensify/Bedrock.git > /dev/null").c_str()));
+
+            // Check out the release tag.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && cd Bedrock && git checkout " + bedrockTagName + "  > /dev/null").c_str()));
+
+            // Build the release.
+            ASSERT_FALSE(system(("cd " + brReleaseDir + " && cd Bedrock && make -j8 bedrock  > /dev/null").c_str()));
+
+            // Save the final product.
+            ASSERT_FALSE(system(("mv " + brReleaseDir + "/Bedrock/bedrock " + prodBedrockName).c_str()));
+        }
+
+        tester = new BedrockClusterTester("db,cache,jobs", prodBedrockName);
+    }
+
+    void teardown() {
+        delete tester;
+    }
```

Comment:
We should probably delete the temp directory here, otherwise this could start to fill up somebody's local disk

Reply:
I left it on purpose because it serves as a sort of cache for running this test multiple times in a row and not needing to rebuild this every time.

On travis/jenkins it doesn't really matter, and it only makes a difference in dev if someone is running this test a lot, which is when caching it is most useful.

Also, it's in /tmp, so it all gets deleted with a reboot anyway. And is uses about 20mb of space, so it's not a ton.